pkg/device/iolimit/utils.go

```diff
@@ -0,0 +1,127 @@
+package iolimit
+
+import (
+	"github.com/openebs/lib-csi/pkg/common/errors"
+	"github.com/openebs/lib-csi/pkg/common/helpers"
+	"io/ioutil"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+const (
+	baseCgroupPath = "/sys/fs/cgroup"
+)
+
+// SetIOLimits sets iops, bps limits for a pod with uid podUid for accessing a device named deviceName
+// provided that the underlying cgroup used for pod namespacing is cgroup2 (cgroup v2)
+func SetIOLimits(request *Request) error {
+	if !helpers.DirExists(baseCgroupPath) {
+		return errors.New(baseCgroupPath + " does not exist")
+	}
+	if err := checkCgroupV2(); err != nil {
+		return err
+	}
+	validRequest, err := validate(request)
+	if err != nil {
+		return err
+	}
+	err = setIOLimits(validRequest)
+	return err
+}
+
+func validate(request *Request) (*ValidRequest, error) {
+	if !helpers.IsValidUUID(request.PodUid) {
+		return nil, errors.New("Expected PodUid in UUID format, Got " + request.PodUid)
+	}
+	podCGPath, err := getPodCGroupPath(request.PodUid, request.ContainerRuntime)
+	if err != nil {
+		return nil, err
+	}
+	ioMaxFile := podCGPath + "/io.max"
+	if !helpers.FileExists(ioMaxFile) {
+		return nil, errors.New("io.max file is not present in pod CGroup")
+	}
+	deviceNumber, err := getDeviceNumber(request.DeviceName)
+	if err != nil {
+		return nil, errors.New("Device Major:Minor numbers could not be obtained")
+	}
+	return &ValidRequest{
+		FilePath:     ioMaxFile,
+		DeviceNumber: deviceNumber,
+		IOMax:      request.IOLimit,
+	}, nil
+}
+
+func getPodCGroupPath(podUid string, cruntime string) (string, error) {
+	switch cruntime {
+	case "containerd":
```

Comment:
how do we get this info? is there a way calller can get what run time we are using or this is going to be an env.

Reply:
caller can get this from env variable and pass same to lib

Reply:
is there a way to make it auto detect, instead of passing that as env? That will make it more generalize lib functionality. May be in the next PR.

Reply:
Even docker runtime now support the cgroupV2 from 20.10 release here is the current adoption status of cgroup v2 in containers https://medium.com/nttlabs/cgroup-v2-596d035be4d7